Cleanup parent fallback logic

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #153275
* __->__ #154006

The `parent` in fallback_node_due_to_unsupported_type is a duplication of `unsupported_output_tensor` logic. remove it. tested that the tests in test_add_complex give same codegen. this fixes an issue in mx that @drisspg was running into.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

eellison committed:
Cleanup parent fallback logic

[ghstack-poisoned]

eellison committed:
Update on "Cleanup parent fallback logic"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

drisspg commented on torch/_inductor/lowering.py:
```diff
@@ -1920,9 +1916,11 @@ def unsupported_input_tensor(t: torch.Tensor, parent=None, node=None):
-    if unsupported_input_tensor(t, parent):
+    if node.target is aten.view.dtype and t.dtype.is_complex:
+        return False
+    if unsupported_input_tensor(t, node):
```
nit: `return unsupported_input_tensor(t, node)`

drisspg commented:
Confirmed that this does indeed help
https://fburl.com/frq9aq3n
Versus:
https://fburl.com/fwx06fv9

Still unable to fuse the cast + swizzle

eellison committed:
Update on "Cleanup parent fallback logic"


The `parent` in fallback_node_due_to_unsupported_type is a duplication of `unsupported_output_tensor` logic. remove it. tested that the tests in test_add_complex give same codegen. this fixes an issue in mx that drisspg was running into.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

eellison committed:
Update on "Cleanup parent fallback logic"


The `parent` in fallback_node_due_to_unsupported_type is a duplication of `unsupported_output_tensor` logic. remove it. tested that the tests in test_add_complex give same codegen. this fixes an issue in mx that drisspg was running into.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

drisspg commented on torch/_inductor/lowering.py:
```diff
@@ -1920,9 +1916,11 @@ def unsupported_input_tensor(t: torch.Tensor, parent=None, node=None):
+def unsupported_output_tensor(t: torch.Tensor, node=None):
     "Do not support writing tensor but can read from it"
-    if unsupported_input_tensor(t, parent):
+    if node is not None and node.target is aten.view.dtype and t.dtype.is_complex:
```
I would love if this function was flipped to `supported_output_tensor` this double negatie is annoying

eellison committed:
Update on "Cleanup parent fallback logic"


The `parent` in fallback_node_due_to_unsupported_type is a duplication of `unsupported_output_tensor` logic. remove it. tested that the tests in test_add_complex give same codegen. this fixes an issue in mx that drisspg was running into.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

eellison commented:
@pytorchbot merge

eellison commented:
cc @desertfire @benjaminglass1 `test_fft_c2c_cuda` only fails for aotinductor. mind taking a look ?

eellison commented on torch/_inductor/lowering.py:
```diff
@@ -1920,9 +1916,11 @@ def unsupported_input_tensor(t: torch.Tensor, parent=None, node=None):
+def unsupported_output_tensor(t: torch.Tensor, node=None):
     "Do not support writing tensor but can read from it"
-    if unsupported_input_tensor(t, parent):
+    if node is not None and node.target is aten.view.dtype and t.dtype.is_complex:
```
Thats strange to me because this is not a literal complex view, if youre changing dtype.. but will update

eellison replied:
@benjaminglass1 the test only fails with aot inductor, so I do think its a legitimate bug even outside of including `torch.ops.prims.convert_element_type.default`

eellison committed:
Update on "Cleanup parent fallback logic"


The `parent` in fallback_node_due_to_unsupported_type is a duplication of `unsupported_output_tensor` logic. remove it. tested that the tests in test_add_complex give same codegen. this fixes an issue in mx that drisspg was running into.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

eellison committed:
Update

[ghstack-poisoned]

eellison committed:
Update on "Cleanup parent fallback logic"


The `parent` in fallback_node_due_to_unsupported_type is a duplication of `unsupported_output_tensor` logic. remove it. tested that the tests in test_add_complex give same codegen. this fixes an issue in mx that drisspg was running into.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

eellison commented:
@pytorchbot merge
